lib/Sema/ConstraintSystem.cpp

```diff
@@ -712,30 +712,54 @@ void ConstraintSystem::recordOpenedTypes(
 
 /// Determine how many levels of argument labels should be removed from the
 /// function type when referencing the given declaration.
-static unsigned getNumRemovedArgumentLabels(ASTContext &ctx, ValueDecl *decl,
+static unsigned getNumRemovedArgumentLabels(TypeChecker &TC, ValueDecl *decl,
                                             bool isCurriedInstanceReference,
                                             FunctionRefKind functionRefKind) {
+  unsigned numParameterLists = 0;
+
+  // This is required because if enum case has a function type and passed
+  // as an argument to other function as a value, it has to be treated as
+  // regular function and all of the labels have to be stripped from its
+  // parameters.
+  //
+  // enum E {
+  //   case foo(a: Int)
+  // }
+  // let bar: [Int] = []
+  // bar.map(E.foo)
+  //
+  // `E.foo` has to act as a regular function type passed as a value.
+  if (!TC.getLangOpts().isSwiftVersion3()) {
+    if (auto *EED = dyn_cast<EnumElementDecl>(decl)) {
```

Comment:
It's safer to have this under the version check, but is it actually necessary?

What I'm wondering is if there might be cases where the Swift 3 behavior is also broken.